server/src/domain/library/library.service.ts

```diff
@@ -0,0 +1,468 @@
+        name: JobName.LIBRARY_SCAN,
+        data: {
+          id: library.id,
+          refreshModifiedFiles: !job.force,
+          refreshAllFiles: job.force ?? false,
+        },
+      });
+    }
+    return true;
+  }
+
+  async handleOfflineRemoval(job: IEntityJob): Promise<boolean> {
+    const assetPagination = usePagination(JOBS_ASSET_PAGINATION_SIZE, (pagination) => {
+      return this.assetRepository.getWith(pagination, WithProperty.IS_OFFLINE, job.id);
+    });
+
+    const assetIds: string[] = [];
+
+    for await (const assets of assetPagination) {
+      for (const asset of assets) {
+        assetIds.push(asset.id);
+      }
+    }
+
+    this.logger.verbose(`Found ${assetIds.length} offline assets to remove`);
+    await this.deleteAssets(assetIds);
+    return true;
+  }
+
+  async handleQueueAssetRefresh(job: ILibraryRefreshJob): Promise<boolean> {
+    const library = await this.repository.get(job.id);
+    if (!library || library.type !== LibraryType.EXTERNAL) {
+      this.logger.warn('Can only refresh external libraries');
+      return false;
+    }
+
+    const user = await this.userRepository.get(library.ownerId);
+    if (!user?.externalPath) {
+      this.logger.warn('User has no external path set, cannot refresh library');
+      return false;
+    }
+
+    this.logger.verbose(`Refreshing library: ${job.id}`);
+    const crawledAssetPaths = (
+      await this.storageRepository.crawl({
+        pathsToCrawl: library.importPaths,
+        exclusionPatterns: library.exclusionPatterns,
+      })
+    )
+      .map(path.normalize)
+      .filter((assetPath) =>
+        // Filter out paths that are not within the user's external path
+        assetPath.match(new RegExp(`^${user.externalPath}`)),
+      );
+
+    this.logger.debug(`Found ${crawledAssetPaths.length} assets when crawling import paths ${library.importPaths}`);
+    const assetsInLibrary = await this.assetRepository.getByLibraryId([job.id]);
+    const offlineAssets = assetsInLibrary.filter((asset) => !crawledAssetPaths.includes(asset.originalPath));
+    this.logger.debug(`${offlineAssets.length} assets in library are not present on disk and will be marked offline`);
+
+    for (const offlineAsset of offlineAssets) {
+      const offlineJobData: IOfflineLibraryFileJob = {
+        id: job.id,
+        assetPath: offlineAsset.originalPath,
+      };
+
+      await this.jobRepository.queue({ name: JobName.LIBRARY_MARK_ASSET_OFFLINE, data: offlineJobData });
+    }
+
+    if (crawledAssetPaths.length > 0) {
+      let filteredPaths: string[] = [];
+      if (job.refreshAllFiles || job.refreshModifiedFiles) {
+        filteredPaths = crawledAssetPaths;
+      } else {
+        const existingPaths = await this.repository.getOnlineAssetPaths(job.id);
+        this.logger.debug(`Found ${existingPaths.length} existing asset(s) in library ${job.id}`);
+
+        filteredPaths = crawledAssetPaths.filter((assetPath) => !existingPaths.includes(assetPath));
```

Comment:
I tested this PR with a library of ~160k files, seems to work pretty well so far. Yet, refreshing the library (only new files) seems to take some time (~10 min on my machine). After observing the debug log, I suspect this statement to be the root cause.

As `existingPaths` is an array, this needs to iterate n*n elements worst-case. I guess using a set for `existingPath` would improve the performance.

Reply:
The reason we're not using sets is because it makes files import in a random order (sets aren't ordered). While the end outcome is the same, I think it's much more predictable for the user that we're importing the newest files first (somewhat like gphotos). Do you have suggestions on improving it in another way?

Checking your debug log, how much time is spent on the crawl vs filtering assets vs actual import?

Reply:
Thanks for your reply, @etnoy . I’m by no ways an expert in JS or TS, but as far as I know, sets guarantee that the iteration order is the same as the insertion order, the same way as arrays do.

> Set objects are collections of values. A value in the set may only occur once; it is unique in the set's collection. You can iterate through the elements of a set in insertion order. The insertion order corresponds to the order in which each element was inserted into the set by the [add()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/add) method successfully (that is, there wasn't an identical element already in the set when add() was called).

see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set

If this doesn’t hold, it’s probably still faster to first convert the `existingPaths` array to a set, as it reduces the filter’s time complexity to O(n) from O(n^2).

Regarding the times, I‘ll provide them later, when I‘m back home again.

Reply:
Yes, the job needs to complete within the minimum amount of time otherwise it assumes it failed and retries it. You can ask for a longer time, etc. Might be better to write more efficient code if possible or break it up into separate jobs.

Reply:
> Thanks for your reply, @etnoy . I’m by no ways an expert in JS or TS, but as far as I know, sets guarantee that the iteration order is the same as the insertion order, the same way as arrays do.
> 
> > Set objects are collections of values. A value in the set may only occur once; it is unique in the set's collection. You can iterate through the elements of a set in insertion order. The insertion order corresponds to the order in which each element was inserted into the set by the [add()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/add) method successfully (that is, there wasn't an identical element already in the set when add() was called).
> 
> see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set
> 
> If this doesn’t hold, it’s probably still faster to first convert the `existingPaths` array to a set, as it reduces the filter’s time complexity to O(n) from O(n^2).
> 
> -Regarding the times, I‘ll provide them later, when I‘m back home again.-
> 
> Edit: Here is the relevant part of the log, this time with less load on the machine:
> 
> ```
> [Nest] 7  - 09/19/2023, 8:03:40 PM VERBOSE [LibraryService] Refreshing library: 2442d4c8-abad-4662-b5d9-0c20787a5e4a
> [Nest] 7  - 09/19/2023, 8:04:00 PM   DEBUG [LibraryService] Found 168416 assets when crawling import paths /photos
> [Nest] 7  - 09/19/2023, 8:05:09 PM   DEBUG [LibraryService] 0 assets in library are not present on disk and will be marked offline
> [Nest] 7  - 09/19/2023, 8:05:10 PM   DEBUG [LibraryService] Found 168416 existing asset(s) in library 2442d4c8-abad-4662-b5d9-0c20787a5e4a
> [Nest] 7  - 09/19/2023, 8:09:53 PM   DEBUG [LibraryService] After db comparison, 0 asset(s) remain to be imported
> Error: Missing lock for job 2. retryJob
>     at Scripts.finishedErrors (/usr/src/app/node_modules/bullmq/dist/cjs/classes/scripts.js:214:24)
>     at Job.moveToFailed (/usr/src/app/node_modules/bullmq/dist/cjs/classes/job.js:414:32)
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async handleFailed (/usr/src/app/node_modules/bullmq/dist/cjs/classes/worker.js:331:21)
>     at async Worker.retryIfFailed (/usr/src/app/node_modules/bullmq/dist/cjs/classes/worker.js:531:24)
> ```
> 
> So crawling ~20s, checking for online paths ~69s, db query ~1s, comparison ~283s
> 
> Also note that there is this "Missing lock" error. After that, the job is apparently restarted and the same output repeats.

awesome, those numbers tell a whole story. I didn't know that a Set in typescript keeps insertion order (it doesn't in Java from where I'm coming). That does sound like an obvious improvement, thanks

Reply:
I did an initial attempt to move to sets, the issue is that they have no way to perform filtering, I see no immediate way to improve this at the moment. Let's keep the discussion going after merge (which is imminent).

Reply:
Thanks for your effort, @etnoy, and especially this great PR!

Feel free to reach out if you need some input.